### PR TITLE
Deletion of Citation

### PR DESCRIPTION
**Pull request by kakhilreddy**

### CODE REVIEW

**Comment by jdamerow** [comment]
Please take care of the code factor complains.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+public class DeleteItemController {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Autowired
+    private ICitationManager citationManager;
+
+    @Autowired
+    private ICitationHelper citationHelper;
+
+    @RequestMapping(value = "/auth/group/{zoteroGroupId}/items/{itemId}/delete")
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java]: just make this a delete request and remove the '/delete' suffix

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    public String storeItem(@ModelAttribute CitationForm form, Authentication authentication, Model model,
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java]: why is this called `storeItem`?

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        List<String> collectionIds = new ArrayList<>();
+        if (form.getCollectionId() != null && !form.getCollectionId().trim().isEmpty()) {
+            collectionIds.add(form.getCollectionId());
+        }
+        try {
+            citationManager.deleteCitation((IUser) authentication.getPrincipal(), zoteroGroupId, collectionIds,
+                    citation);
+        } catch (ZoteroConnectionException | GroupDoesNotExistException | ZoteroHttpStatusException
+                | ZoteroItemDeletionFailedException e) {
+            e.printStackTrace();
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java]: not how you would handle an exception.

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp [hunk 4]
```diff
@@ -271,6 +273,22 @@
...
+  <div class="modal-dialog" role="document">
+    <div class="modal-content">
+     
+      <div class="modal-body">
+    Are you sure you want to delete Citation?
+      </div>
+      <div class="modal-footer">
+        <button type="button" class="btn btn-secondary" data-dismiss="modal">No</button>
+        <a class="btn btn-primary"
+		href="<c:url value="/auth/group/${zoteroGroupId}/items/${citation.key}/delete"/>" > Yes</a>
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp]: requests that change the data on the server should never be GET requests.

File: citesphere/pom.xml [hunk 5]
```diff
@@ -18,7 +18,7 @@
         <spring-data.version>Lovelace-RELEASE</spring-data.version>
         <javers.version>5.0.0</javers.version>
         <spring.kafka.version>2.2.6.RELEASE</spring.kafka.version>
-        <spring-social-zotero.version>0.1</spring-social-zotero.version>
+        <spring-social-zotero.version>0.2-SNAPSHOT</spring-social-zotero.version>
```
Comment (jdamerow) [inline citesphere/pom.xml]: the new version has been published, so you should be able to use 0.2

Comment (jdamerow) [reply citesphere/pom.xml]: now it should be 0.4

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/exceptions/ZoteroItemDeletionFailedException.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
+package edu.asu.diging.citesphere.core.exceptions;
+
+import org.springframework.social.zotero.api.ItemCreationResponse;
+
+public class ZoteroItemDeletionFailedException extends Exception {
+    private ItemCreationResponse response;
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/exceptions/ZoteroItemDeletionFailedException.java]: why does a deletion failed exception have an `ItemCreationResponse` object?

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java [hunk 7]
```diff
@@ -314,14 +342,31 @@ private PageRequest createPageRequest(IUser user, int page, String sortBy, ICita
...
+    @Override
+    public boolean deleteCitation(IUser user, String groupId, ICitation citation) throws ZoteroConnectionException,
+            GroupDoesNotExistException, ZoteroHttpStatusException, ZoteroItemDeletionFailedException {
+        Optional<CitationGroup> groupOptional = groupRepository.findById(new Long(groupId));
+        if (!groupOptional.isPresent()) {
+            throw new GroupDoesNotExistException();
+        }
+
+        zoteroManager.deleteCitation(user, groupId, citation);
+        citationRepository.delete((Citation) citation);
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java]: this will probably fail if citation is referenced from any page request

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/zotero/impl/ZoteroConnector.java [hunk 8]
```diff
@@ -242,11 +244,25 @@ private Zotero getApi(IUser user) {
...
     private ZoteroHttpStatusException createException(HttpStatus status, Exception cause) {
         if (status == HttpStatus.FORBIDDEN) {
             return new AccessForbiddenException(cause);
         }
         return new ZoteroHttpStatusException(cause);
     }
+
+    @Override
+    public void deleteItem(IUser user, String groupId, String citationKey, Long citationVersion)
+            throws ZoteroConnectionException, ZoteroHttpStatusException, ZoteroItemDeletionFailedException {
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/zotero/impl/ZoteroConnector.java]: where is `ZoteroItemDeletionFailedException` thrown?

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp [hunk 9]
```diff
@@ -250,129 +257,228 @@
...
+		<div class="modal-content">
+
+			<div class="modal-body">Are you sure you want to delete
+				Citation?</div>
+			<div class="modal-footer">
+
+				<button type="button" class="btn btn-secondary" data-dismiss="modal">No</button>
+				<button id="delete" type="button" class="btn btn-primary">Yes</button>
+				<%-- <a class="btn btn-primary"
+		href="<c:url value="/auth/group/${zoteroGroupId}/items/${citation.key}/delete"/>" > Yes</a>  --%>
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp]: delete if you don't need this

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp [hunk 10]
```diff
@@ -250,129 +257,228 @@
...
+					$("#parsed-ref-reference-label").text(
+							$(clicked).attr('data-reference-label'));
+					$("#parsed-ref-citation-id").text(
+							$(clicked).attr('data-citation-id'));
+					$("#parsed-ref-ref-raw").text(
+							$(clicked).attr('data-ref-raw'));
+
+					$("#parsed-ref-modal").modal();
+				});
+
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp]: why is this indented so much?

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java (hunk 7 above)
Comment (kakhilreddy) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:365]: what should i do about it

Comment (jdamerow) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:365]: delete all the references to it first

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java [hunk 11]
```diff
@@ -310,18 +338,42 @@ private PageRequest createPageRequest(IUser user, int page, String sortBy, ICita
...
+        Optional<CitationGroup> groupOptional = groupRepository.findById(new Long(groupId));
+        if (!groupOptional.isPresent()) {
+            throw new GroupDoesNotExistException();
+        }
+
+        zoteroManager.deleteCitation(user, groupId, citation);
+
+        List<PageRequest> requests = pageRequestRepository.findPageRequestByKey(citation.getKey());
+        for (PageRequest request : requests) {
+            boolean delete = request.getCitations().remove(citation);
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:368]: why do you store the return value if you don't use it?

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/zotero/impl/ZoteroConnector.java [hunk 12]
```diff
@@ -242,11 +244,27 @@ private Zotero getApi(IUser user) {
...
+        Zotero zotero = getApi(user);
+        if (!zotero.getGroupsOperations().deleteItem(groupId, citationKey, citationVersion)) {
+            throw new ZoteroItemDeletionFailedException("Deletion of Citation failed");
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/zotero/impl/ZoteroConnector.java:260]: what's the difference between ZoteroItemDeletionFailedException and ZoteroHttpStatusException?

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/repository/cache/PageRequestRepository.java [hunk 13]
```diff
@@ -13,11 +13,15 @@
...
-    @Query(value="SELECT DISTINCT pr from PageRequest pr LEFT JOIN FETCH pr.citations c LEFT JOIN FETCH c.authors a WHERE pr.objectId = ?2 AND pr.user = ?1 AND pr.pageNumber = ?3 AND pr.zoteroObjectType = ?4")
+    @Query(value = "SELECT DISTINCT pr from PageRequest pr LEFT JOIN FETCH pr.citations c LEFT JOIN FETCH c.authors a WHERE pr.objectId = ?2 AND pr.user = ?1 AND pr.pageNumber = ?3 AND pr.zoteroObjectType = ?4")
 //    @EntityGraph(value="requestsWithFullCitations")
-    List<PageRequest> findPageRequestWithCitations(IUser user, String objectId, int page, ZoteroObjectType zoteroObjectType);
-} 
+    List<PageRequest> findPageRequestWithCitations(IUser user, String objectId, int page,
+            ZoteroObjectType zoteroObjectType);
 
+    @Query(value = "SELECT PageRequest.* FROM PageRequest INNER JOIN PageRequest_Citation ON PageRequest.id = PageRequest_Citation.PageRequest_id where PageRequest_Citation.citations_citationKey = ?1", nativeQuery = true)
+    List<PageRequest> findPageRequestByKey(String citationKey);
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/repository/cache/PageRequestRepository.java:26]: have you tried using automatic query generation? should be easy to do and easier to maintain.

Comment (kakhilreddy) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/repository/cache/PageRequestRepository.java:26]: i tried but it was not working.
So i used native query

Comment (jdamerow) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/repository/cache/PageRequestRepository.java:26]: what did you try?

Comment (kakhilreddy) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/repository/cache/PageRequestRepository.java:26]: @EntityGraph(value = "requestsWithFullCitations", type = EntityGraphType.LOAD)
    List<PageRequest> findByKey(String key);

I am not sure how to search based on Entity graph

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java [hunk 14]
```diff
@@ -310,18 +337,41 @@ private PageRequest createPageRequest(IUser user, int page, String sortBy, ICita
...
-    
+
     @Override
     public List<String> getValidCreatorTypes(IUser user, ItemType itemType) {
         return zoteroManager.getValidCreatorTypes(user, itemType);
     }
-}
+
+    @Override
+    public void deleteCitation(IUser user, String groupId, ICitation citation)
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:356]: needs a test case

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/zotero/impl/ZoteroConnector.java [hunk 15]
```diff
@@ -242,11 +244,25 @@ private Zotero getApi(IUser user) {
...
+
     private ZoteroHttpStatusException createException(HttpStatus status, Exception cause) {
         if (status == HttpStatus.FORBIDDEN) {
             return new AccessForbiddenException(cause);
         }
         return new ZoteroHttpStatusException(cause);
     }
+
+    @Override
+    public void deleteItem(IUser user, String groupId, String citationKey, Long citationVersion)
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/zotero/impl/ZoteroConnector.java:256]: needs a test case

**Comment by jdamerow** [comment]
Jenkins, test this please.

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp [hunk 16]
```diff
@@ -30,6 +35,8 @@
 	<a
 		href="<c:url value="/auth/group/${zoteroGroupId}/items/${citation.key}/edit" />"><i
 		class="far fa-edit" title="Edit"></i></a> &nbsp;&nbsp; <a
+		data-toggle="modal" data-target="#deleteModal" href=""><i
+		class="fas fa-trash" title="Sync Citation"></i></a> &nbsp;&nbsp; <a
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:39]: why does the title say "Sync Citation"?

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp [hunk 17]
```diff
@@ -250,129 +257,222 @@
...
-    </div><!-- /.modal-content -->
-  </div><!-- /.modal-dialog -->
-</div><!-- /.modal -->
+<div class="modal fade" id="deleteModal" tabindex="-1" role="dialog"
+	aria-labelledby="deleteModalLabel" aria-hidden="true">
+	<div class="modal-dialog" role="document">
+		<div class="modal-content">
+
+			<div class="modal-body">Are you sure you want to delete
+				Citation?</div>
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:283]: There should be a title for the modal

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp [hunk 18]
```diff
@@ -250,129 +257,222 @@
...
+					$("#parsed-ref-reference-label").text(
+							$(clicked).attr('data-reference-label'));
+					$("#parsed-ref-citation-id").text(
+							$(clicked).attr('data-citation-id'));
+					$("#parsed-ref-ref-raw").text(
+							$(clicked).attr('data-ref-raw'));
+
+					$("#parsed-ref-modal").modal();
+				});
+
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:430]: why did you change the indentation?

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp [hunk 19]
```diff
@@ -250,129 +257,222 @@
...
+	    $('#deleteModal').modal("hide");
+	    $('#loadModal').modal("show");
+	    $.ajax({
+		    url : '<c:url value="/auth/group/${zoteroGroupId}/items/${citation.key}"/>'+ "?${_csrf.parameterName}=${_csrf.token}",
+		    method : 'DELETE',
+	        'success' : function(data) {
+				
+
+                $('#loadModal').modal("hide");
+                $.notify('<i class="fas fa-check-circle"></i> Citation successfully deleted!',
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:443]: the notification should stay there until the new page is loaded and it should say
"Citation successfully delete. You will be redirected to the citation list page."

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp [hunk 20]
```diff
@@ -250,129 +257,222 @@
...
+
+<div class="modal fade" id="loadModal" tabindex="-1" role="dialog"
+	aria-labelledby="loadModalLabal" aria-hidden="true">
+	<div class="modal-dialog" role="document">
+		<div class="modal-content">
+
+			<div class="modal-body">
+
+				<div>
+					<span><i class="fas fa-spinner text-center"></i> Deleting...</span>
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:302]: the spinner should spin

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp (hunk 18 above)
Comment (kakhilreddy) [reply citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:430]: since they were coming too far in github , i manually added spaces instead of tabs

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp (hunk 20 above)
Comment (kakhilreddy) [reply citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:302]: Before we discussed that the spinner icon would be fine.

Comment (jdamerow) [reply citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:302]: maybe I misunderstood you, but the spinner should spin. It should be just one more css class, so should be easy to do.

**Comment by jdamerow** [comment]
Jenkins, test this please.

File: citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java [hunk 21]
```diff
@@ -37,217 +40,273 @@
...
     }
-    
+
+    @Test
+    public void test_deleteCitation_success()
+            throws ZoteroConnectionException, GroupDoesNotExistException, ZoteroHttpStatusException {
+        /*
+         * ICitation deleteCitation = new Citation();
+         * deleteCitation.setKey(EXISTING_ID); deleteCitation.setVersion(new Long(2));
+         */
```
Comment (jdamerow) [inline citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java:172]: remove if not needed

File: citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java [hunk 22]
```diff
@@ -37,217 +40,273 @@
...
+        PageRequest request = new PageRequest();
+        request.setCitations(citations);
+        requests.add(request);
+
+        managerToTest.deleteCitation(user, GROUP_ID, existingCitation);
+
+        Mockito.when(groupRepository.findById(new Long(GROUP_ID))).thenReturn(optionalGroup);
+        Mockito.verify(zoteroManager).deleteCitation(user, GROUP_ID, existingCitation);
+        Mockito.verify(citationRepository).delete((Citation) existingCitation);
+        Mockito.when(pageRequestRepository.findByCitations(existingCitation)).thenReturn(requests);
```
Comment (jdamerow) [inline citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java:193]: why is that here? you already called the manager.

Comment (jdamerow) [reply citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java:193]: also, you want to make sure that the citation has been removed from the page request citations list

**Comment by jdamerow** [comment]
when a deletion fails because the citation wasn't the latest version, it should tell that to the user and ask if they want to sync the citation.

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp (hunk 9 above)
Comment (jdamerow) [reply citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:294]: ?

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp [hunk 23]
```diff
@@ -250,129 +259,238 @@
...
-    </div><!-- /.modal-content -->
-  </div><!-- /.modal-dialog -->
-</div><!-- /.modal -->
+<div class="modal fade" id="deleteModal" tabindex="-1" role="dialog"
+	aria-labelledby="deleteModalLabel" aria-hidden="true">
+	<div class="modal-dialog" role="document">
+		<div class="modal-content">
+	<div class="modal-header">
+			
+				<h4 class="modal-title">Warning</h4>
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:285]: title should be "Delete Citation"

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+
+    @RequestMapping(value = "/auth/group/{zoteroGroupId}/items/{itemId}", method = RequestMethod.DELETE)
+    public ResponseEntity<String> deleteItem(@PathVariable("zoteroGroupId") String zoteroGroupId,
+            @PathVariable("itemId") String itemId, Authentication authentication, Model model)
+            throws ZoteroConnectionException, GroupDoesNotExistException, CannotFindCitationException,
+            ZoteroHttpStatusException {
+        ICitation citation = citationManager.getCitation((IUser) authentication.getPrincipal(), zoteroGroupId, itemId);
+        try {
+            citationManager.deleteCitation((IUser) authentication.getPrincipal(), zoteroGroupId, citation);
+        } catch (CitationIsOutdatedException e) {
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java:50]: what about all the other exceptions? They should be handled appropriately.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+            return new ResponseEntity<>(
+                    "Deletion Failed. Citation isn't the latest version. Please sync the citation and try again",
+                    HttpStatus.INTERNAL_SERVER_ERROR);
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java:53]: there is probably a better error code than 500. also, the text should be in the jsp page, not in the response.

Comment (kakhilreddy) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java:53]: Its an AJAX request, so i sent a response message

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java (hunk 24 above)
Comment (kakhilreddy) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java:50]: I wrote based on EditItemController. In that class, only CitationIsOutdatedException was handled.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java (hunk 25 above)
Comment (jdamerow) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java:53]: even in an ajax request you can write error messages in the page not the code

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java (hunk 24 above)
Comment (jdamerow) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java:50]: not an excuse

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java (hunk 25 above)
Comment (kakhilreddy) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java:53]: Im confused with this. 
Do i need to handle errors with model attributes.

Comment (jdamerow) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java:53]: no, your javascript code should show different things according to the error code returned.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+    private ICitationManager citationManager;
+
+    @Autowired
+    private ICitationHelper citationHelper;
+
+    @RequestMapping(value = "/auth/group/{zoteroGroupId}/items/{itemId}", method = RequestMethod.DELETE)
+    public ResponseEntity<String> deleteItem(@PathVariable("zoteroGroupId") String zoteroGroupId,
+            @PathVariable("itemId") String itemId, Authentication authentication, Model model)
+
+    {
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteItemController.java:47]: please follow our style guide

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp (hunk 17 above)
Comment (jdamerow) [reply citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:288]: Should be: Are you sure you want to delete this citation?

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp [hunk 27]
```diff
@@ -250,129 +259,241 @@
...
 			var clicked = e.target;
 			$("#ref-unparsed").text($(clicked).text());
-			$("#parsed-ref-authors").text($(clicked).attr('data-authors'));
-			$("#parsed-ref-title").text($(clicked).attr('data-title'));
-			$("#parsed-ref-year").text($(clicked).attr('data-year'));
-			var identifier = $(clicked).attr('data-identifier');
-			var identType = $(clicked).attr('data-identifier-type');
-			if (identType != undefined && identType != "") {
-				identifier = identifier + " (" + identType + ")";
+			$("#parsed-ref-authors").text(
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:404]: why is here a newline?

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp (hunk 10 above)
Comment (jdamerow) [reply citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:435]: fix indentation!

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp [hunk 28]
```diff
@@ -250,129 +259,241 @@
...
+						
+						window.location.href = '<c:url value="/auth/group/${zoteroGroupId}/items"/>'
+					}, 1);
+
+					},
+			'error' : function(data) {
+				console.log(data);
+			
+			if (data.status == 412) {
+				msg = 'Citation is outdated. Please Sync the citation and try again';
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:473]: "sync" needs to be lowercase. Also, let's make that a link to click that syncs the citation.

Comment (jdamerow) [reply citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:473]: Also, period is missing at the end of the sentence (same for all the other messages).

File: citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java [hunk 29]
```diff
@@ -12,6 +12,8 @@
 import org.mockito.InjectMocks;
 import org.mockito.Mock;
 import org.mockito.Mockito;
+import static org.mockito.Mockito.doThrow;
```
Comment (jdamerow) [inline citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java:15]: please use `Mockito.` prefix like all the other calls to mockito to be consistent.

File: citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java [hunk 30]
```diff
@@ -37,217 +40,293 @@
...
+        Mockito.when(citationRepository.findById(existingCitation.getKey())).thenReturn(Optional.of(existingCitation));
+        Mockito.when(groupRepository.findById(new Long(GROUP_ID))).thenReturn(optionalGroup);
+        Mockito.when(pageRequestRepository.findByCitations(existingCitation)).thenReturn(requests);
+
+        managerToTest.deleteCitation(user, GROUP_ID, existingCitation);
+
+        Mockito.verify(zoteroManager).deleteCitation(user, GROUP_ID, existingCitation);
+        Mockito.verify(citationRepository).delete((Citation) existingCitation);
+
+        assert (requests.get(0).getCitations().size() == 0);
```
Comment (jdamerow) [inline citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java:211]: pleae use junit's assert

File: citesphere/src/test/java/edu/asu/diging/citesphere/core/zotero/impl/ZoteroConnectorTest.java [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+    @Mock
+    private GroupsOperations operations;
+
+    private static String CITATION_KEY = "12";
+
+    private final String GROUP_ID = "12";
+
+    private IUser user;
+
+    Long currentVersion = new Long(1);
```
Comment (jdamerow) [inline citesphere/src/test/java/edu/asu/diging/citesphere/core/zotero/impl/ZoteroConnectorTest.java:47]: why is that not private?

File: citesphere/src/test/java/edu/asu/diging/citesphere/core/zotero/impl/ZoteroConnectorTest.java [hunk 32]
```diff
@@ -0,0 +1,83 @@
...
+    public void setUp() throws ZoteroHttpStatusException {
+        MockitoAnnotations.initMocks(this);
+
+        user = new User();
+        user.setUsername("USERNAME");
+
+    }
+
+    @Test
+    public void test_deleteItem_success() throws ZoteroConnectionException, ZoteroHttpStatusException {
```
Comment (jdamerow) [inline citesphere/src/test/java/edu/asu/diging/citesphere/core/zotero/impl/ZoteroConnectorTest.java:71]: what about the two cases when it throws exceptions?

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp [hunk 33]
```diff
@@ -250,129 +259,242 @@
...
+			if (data.status == 412) {
+				msg =  '<a href="<c:url value="/auth/group/${zoteroGroupId}/items/${citation.key}/sync" />">Citation is outdated. Please sync the citation and try again. </a>'
+				;
+				
+			} else if (data.status == 404) {
+				msg = 'Group or Citation not found';
+			} else {
+				msg = 'Citation deletion failed. Please try again.';
+				
+			}
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:481]: indentation

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp [hunk 34]
```diff
@@ -250,129 +259,242 @@
...
+						
+						window.location.href = '<c:url value="/auth/group/${zoteroGroupId}/items"/>'
+					}, 1);
+
+					},
+			'error' : function(data) {
+				console.log(data);
+			
+			if (data.status == 412) {
+				msg =  '<a href="<c:url value="/auth/group/${zoteroGroupId}/items/${citation.key}/sync" />">Citation is outdated. Please sync the citation and try again. </a>'
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:473]: only "sync the citation" should be a link

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp [hunk 35]
```diff
@@ -250,129 +259,242 @@
...
+				;
+				
+			} else if (data.status == 404) {
+				msg = 'Group or Citation not found';
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.jsp:477]: period is missing

File: citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java (hunk 29 above)
Comment (jdamerow) [reply citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java:15]: ?

File: citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java [hunk 36]
```diff
@@ -37,217 +40,303 @@
...
+        Mockito.when(citationRepository.findById(existingCitation.getKey())).thenReturn(Optional.of(existingCitation));
+        Mockito.when(groupRepository.findById(new Long(GROUP_ID))).thenReturn(optionalGroup);
+        Mockito.when(pageRequestRepository.findByCitations(existingCitation)).thenReturn(requests);
+
+        managerToTest.deleteCitation(user, GROUP_ID, existingCitation);
+
+        Mockito.verify(zoteroManager).deleteCitation(user, GROUP_ID, existingCitation);
+        Mockito.verify(citationRepository).delete((Citation) existingCitation);
+
+        Assert.assertTrue(requests.get(0).getCitations().size() == 0);
```
Comment (jdamerow) [inline citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java:220]: let's use `assertEquals` so the if this fails the error message will state what was expected and what was the value.

File: citesphere/src/test/java/edu/asu/diging/citesphere/core/zotero/impl/ZoteroConnectorTest.java (hunk 32 above)
Comment (jdamerow) [reply citesphere/src/test/java/edu/asu/diging/citesphere/core/zotero/impl/ZoteroConnectorTest.java:71]: ?

**Comment by jdamerow** [comment]
Please resolve conflicts.
